File: services/stream-processor/health.go
package main

import (
	"context"
	"encoding/json"
	"log"
	"net/http"
	"time"

	"github.com/prometheus/client_golang/prometheus/promhttp"
)

func startHealthServer(port string, dedup *deduplicator, influx *influxSink, kio *kafkaIO) {
	mux := http.NewServeMux()

	mux.HandleFunc("/live", func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusOK)
		_, _ = w.Write([]byte("ok"))
	})

	mux.HandleFunc("/ready", func(w http.ResponseWriter, r *http.Request) {
		ctx, cancel := context.WithTimeout(r.Context(), 2*time.Second)
		defer cancel()

		redisOK := dedup.ping(ctx) == nil
		influxState := influx.breakerState()
		kafkaState := kio.breakerState()
		ready := redisOK && influxState != "OPEN" && kafkaState != "OPEN"

		w.Header().Set("Content-Type", "application/json")
		if !ready {
			w.WriteHeader(http.StatusServiceUnavailable)
		}
		_ = json.NewEncoder(w).Encode(map[string]any{
			"ready":                    ready,
			"redis_connected":          redisOK,
			"influxdb_circuit_breaker": influxState,
			"kafka_circuit_breaker":    kafkaState,
		})
	})

	mux.Handle("/metrics", promhttp.Handler())

	go func() {
		// ListenAndServe only returns once the listener stops; nothing in
		// this service ever calls Shutdown/Close on it, so any return here
		// is unexpected (most likely a bind failure, e.g. the port already
		// in use) and worth surfacing loudly rather than leaving /live and
		// /ready silently unreachable for the rest of the process's life.
		if err := http.ListenAndServe(":"+port, mux); err != nil { //nolint:gosec // internal-only health/metrics endpoint
			log.Printf("stream-processor: health/metrics server on :%s stopped: %v", port, err)
		}
	}()
}

File: services/stream-processor/health.py
"""HTTP health/metrics server, the Python port of health.go. Same three
endpoints, same semantics: /live never depends on a dependency, /ready
reflects real Redis+InfluxDB+Kafka state, /metrics is the Prometheus scrape
endpoint.
"""

from __future__ import annotations

import json
import logging
import threading
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer

from prometheus_client import CONTENT_TYPE_LATEST, generate_latest

from dedup import Deduplicator
from influx import InfluxSink
from kafka_io import KafkaIO

_logger = logging.getLogger("stream-processor")


def _make_handler(dedup: Deduplicator, influx: InfluxSink, kio: KafkaIO) -> type[BaseHTTPRequestHandler]:
    class Handler(BaseHTTPRequestHandler):
        def log_message(self, fmt: str, *args) -> None:  # noqa: A002 - silence stdlib's default access log
            pass

        def do_GET(self) -> None:
            if self.path == "/live":
                self.send_response(200)
                self.end_headers()
                self.wfile.write(b"ok")
                return

            if self.path == "/ready":
                redis_ok = dedup.ping()
                influx_state = influx.breaker_state()
                kafka_state = kio.breaker_state()
                ready = redis_ok and influx_state != "OPEN" and kafka_state != "OPEN"
                body = json.dumps(
                    {
                        "ready": ready,
                        "redis_connected": redis_ok,
                        "influxdb_circuit_breaker": influx_state,
                        "kafka_circuit_breaker": kafka_state,
                    }
                ).encode("utf-8")
                self.send_response(200 if ready else 503)
                self.send_header("Content-Type", "application/json")
                self.end_headers()
                self.wfile.write(body)
                return

            if self.path == "/metrics":
                body = generate_latest()
                self.send_response(200)
                self.send_header("Content-Type", CONTENT_TYPE_LATEST)
                self.end_headers()
                self.wfile.write(body)
                return

            self.send_response(404)
            self.end_headers()

    return Handler


def start_health_server(port: str, dedup: Deduplicator, influx: InfluxSink, kio: KafkaIO) -> ThreadingHTTPServer | None:
    """Starts the health/metrics server in a background thread. A bind
    failure (e.g. the port already in use) is logged loudly rather than
    silently leaving /live and /ready unreachable for the process's whole
    lifetime — the exact fix a pre-GitHub audit made to health.go, ported
    here rather than reintroducing the original bug."""
    try:
        server = ThreadingHTTPServer(("", int(port)), _make_handler(dedup, influx, kio))
    except OSError as exc:
        _logger.error("stream-processor: health/metrics server on :%s stopped: %s", port, exc)
        return None
    thread = threading.Thread(target=server.serve_forever, daemon=True)
    thread.start()
    return server
